openstack/orchestration/v1/resourcetypes/results.go

```diff
@@ -0,0 +1,163 @@
+package resourcetypes
+
+import (
+	"github.com/gophercloud/gophercloud"
+)
+
+// ResourceTypeList contains the result of listing the available resource types.
+// It includes both a list of resource types and a map of their descriptions.
+// The descriptions will be empty unless they were requested from the server.
+type ResourceTypeList struct {
```

Comment:
This doesn't look to be modeled like the API response. In the test fixtures, you have:

```json
{
    "resource_types": [
        {
            "description": "A Nova Server",
	    "resource_type": "OS::Nova::Server"
        },
        {
            "description": "A Heat Stack",
	    "resource_type": "OS::Heat::Stack"
        }
    ]
}
```

So a modeled response would be:

```go
type ResourceTypeResult struct {
  ResourceType string `json:"resource_type"`
  Description string `json:"description"`
}
```

And then the result of `Extract` would be `[]ResourceTypeResult`.

I'm not a huge fan of the name `ResourceTypeResult`, but it's better than naming it `ResourceType` while also having `ResourceType` as a field name within the struct. Alternatively,

```go
type ResourceTypeResult struct {
  Name string `json:"resource_type"`
  Description string `json:"description"`
}
```

Which breaks convention, but might be worth an exception in this case.

Reply:
The problem is that the modelled result isn't very useful in practice. What people are mostly going to do with this data is either iterate over the whole list somehow (possibly using the descriptions, or possibly not) or test for whether a particular resource type is present.

Maybe the best thing to return would be a `map[string]ResourceTypeResult` so that the name appears as both the key and within the result struct, but the caller gets the data returned in a way that they can use without the weird doubling up in my ResourceTypeList type. Would that work for you? Or is the priority on exactly modelling the API output over usability of the result? (FWIW I probably wrote the original API, and would change it now if I could ;)

Reply:
>  Or is the priority on exactly modelling the API output over usability of the result? 

Yes, this.

I totally understand that the "raw" result isn't useful - we've run into a few other occurrences of "it sounded like a good idea at the time" - but Gophercloud should model the API result as close to the response body as possible. This helps with long-term maintenance of the project because we're able to work strictly with the information the service is sending to us rather than having to ping someone external with historical knowledge.

It would be possible to add a helper function to [gophercloud/utils](https://github.com/gophercloud/utils) which could act as a wrapper and return a more useful result. The utils repo an open repository for most anything that makes Gophercloud more usable, allowing the main Gophercloud repo to be strictly API modeling.